app/models/patient.rb

```diff
@@ -740,23 +740,54 @@ def self.order_by_risk(asc: true)
   end
 
   # Check for potential duplicate records. Duplicate criteria is as follows:
-  # - matching first name, last name, sex, and DoB
+  # - matching values of first name and last name. Optionally sex and DoB can also match for more detailed messaging
+  # e.g. Jon Smith M null would be a duplicate of Jon Smith M 1/1/2000
   # OR
   # - matching state/local id
   def self.duplicate_data(first_name, last_name, sex, date_of_birth, user_defined_id_statelocal)
-    dup_info = where('first_name = ?', first_name)
-               .where('last_name = ?', last_name)
-               .where('sex = ?', sex)
-               .where('date_of_birth = ?', date_of_birth)
+    # Track which matches have occurred
+    duplicate_field_data = []
 
+    # check for a duplicate state/local id
     dup_statelocal_id = where('user_defined_id_statelocal = ?', user_defined_id_statelocal&.to_s&.strip)
+    duplicate_field_data << { count: dup_statelocal_id.size, fields: ['State/Local ID'] } if dup_statelocal_id.present?
 
-    # Get fields that have matching values
-    duplicate_field_data = []
-    duplicate_field_data << { count: dup_info.count, fields: ['First Name', 'Last Name', 'Sex', 'Date of Birth'] } if dup_info.present?
-    duplicate_field_data << { count: dup_statelocal_id.count, fields: ['State/Local ID'] } if dup_statelocal_id.present?
+    # if first_name or last_name is null skip duplicate detection
+    return { is_duplicate: duplicate_field_data.any?, duplicate_field_data: duplicate_field_data } if first_name.nil? || last_name.nil?
+
+    fn_ln_sex_dob_matches = 0
+    fn_ln_sex_matches = 0
+    fn_ln_dob_matches = 0
+    fn_ln_matches = 0
+
+    # Determine which type of duplicate exists
+    where(first_name: first_name, last_name: last_name).pluck(*%i[sex date_of_birth]).each do |(s, dob)|
+      dob = dob&.strftime('%F')
```

Comment:
Thoughts on moving this string format into the query itself? 

`where(first_name: first_name, last_name: last_name).select('sex', "DATE_FORMAT(date_of_birth, '%Y-%m-%d') date_of_birth").pluck(:sex, :date_of_birth)`

Reply:
So I think when `.select` and `.pluck` are being chained like this, the `select` does nothing. Here's an example you can run in the console to verify this: `Patient.limit(10).select(:id).pluck(:first_name)`

Is the goal of this to just return `dob` as a string instead of a `date` from SQL? If so, I think we can write:
```
.pluck(ActiveRecord::Base.sanitize_sql(Arel.sql('CAST(date_of_birth AS CHAR)')))
```

Reply:
You're right ..... can also do this 

```ruby
    where(first_name: patient[:first_name], last_name: patient[:last_name])
      .select('sex', "DATE_FORMAT(date_of_birth, '%Y-%m-%d') date_of_birth").each do |found|
        # If the sex isn't nil and doesn't match it is not a duplicate. Same for DoB.
        if (patient[:sex].present? && found.sex.present? && patient[:sex] != found.sex) ||
           (date_of_birth.present? && found.date_of_birth.present? && date_of_birth != found.date_of_birth.to_s)
          next
        end

        # Check for duplicates
        if patient[:sex].present? && patient[:sex] == found.sex && patient[:date_of_birth].present? && date_of_birth == found.date_of_birth.to_s
          fn_ln_sex_dob_matches += 1
        elsif patient[:sex].present? && patient[:sex] == found.sex
          fn_ln_sex_matches += 1
        elsif patient[:date_of_birth].present? && patient[:date_of_birth] == found.date_of_birth.to_s
          fn_ln_dob_matches += 1
        else
          fn_ln_matches += 1
        end
      end
```

Reply:
BTW I benchmarked formatting with DATE_FORMAT vs pluck and strftime and select formatting is about 8x faster